### ISSUE TITLE
[Feature request] Segmentation head

### ISSUE THREAD

**Issue by johnlockejrr** [ArgoHA/custom_d_fine#36]
Hi there! Sorry to open an issue for no issue at all.
I found this repo yesterday being forked by an acquaintance. I don't want to praise or anything but just: wow!
YOLO detection models are far behind.

Is there any chance the model to have in the future segmentation (like polygon masks not just or no bounding boxes)?
That would be amazing!

🔧 Changes Needed:
1. Model Architecture:
 * Replace bbox head with polygon/segmentation head (or add not replace!)
 * Variable output length - polygons have different numbers of points
 * New loss functions - polygon IoU, Chamfer distance, etc.
 * Output format - [x1, y1, x2, y2, ..., xn, yn] instead of [x1, y1, x2, y2]
 
2. Data Pipeline:
 * YOLO segmentation format - class x1 y1 x2 y2 ... xn yn (normalized)
 * Variable-length handling - different polygons have different point counts
 * Augmentation updates - transform all polygon points correctly

3. Training Changes:
 * New loss functions - polygon-specific losses
 * No pretrained weights - would need to train from scratch
 * Memory management - variable-length outputs are memory intensive

Thank you for yor work! 🥇

**Comment by johnlockejrr** [comment]
Fingers crossed, dude! 👍

**Comment by ArgoHA** [comment]
Rf-detr skipped some lines, interesting, thanks for sharing

**Comment by johnlockejrr** [comment]
Anyway, I had much hopes for `rf-detr` but at least in my case is just a YOLO harder to train and that's all, even worse.

**Comment by ArgoHA** [comment]
Hey @johnlockejrr I fixed a denoising bug, now segmentation works correctly. I am also working on another small improvement, but currently masks will already be training fine for your task

**Comment by ArgoHA** [comment]
There is at least 1 param added under the "train" - "mask_batch_size: 1000"
I would also recommend trying keep_ratio: False for a baseline. Please check current config_.yaml in the repo to see if there are other new fields

**Comment by johnlockejrr** [comment]
So this config should be safe to try?

```
project_name: sam_44_mss # for wandb
exp_name: sam_44_mss # experiment name

exp: ${exp_name}_${now_dir}

model_name: s # model size (n, s, m, l, x)
task: segment

train:
  ### Paths ###
  root: /home/incognito/D-FINE-seg # project root with dataset and outputs
  pretrained_dataset: coco # coco, obj2coco
  pretrained_model_path: pretrained/dfine_${model_name}_${train.pretrained_dataset}.pth # dfine_m_obj2coco.pth

  data_path: ${train.root}/datasets/sam_44_mss # path to dataset
  path_to_test_data: ${train.root}/images  # path to test set, used in infer script
  path_to_save: ${train.root}/output/models/${exp} # where to save output

  debug_img_path: ${train.root}/output/debug_images
  eval_preds_path: ${train.root}/output/eval_preds
  bench_img_path: ${train.root}/output/bench_imgs
  infer_path: ${train.root}/output/infer

  ### Configs ###
  use_wandb: False
  device: cuda
  label_to_name: # dataset's classes
    0: textline

  use_one_class: False

  ### Distributed Training (DDP) ###
  ddp:
    enabled: False  # set to true to use multi-GPU
    n_gpus: 2  # how many GPUs / processes per node for torchrun

  decision_metrics:
  - f1
  - mAP_50

  img_size: [896, 640] # (h, w)
  keep_ratio: False # image aspect ratio, if True - image will be padded
  to_visualize_eval: True # save images with gt and preds
  debug_img_processing: True # save images after preprocessing

  amp_enabled: True # use automatic mixed precision
  clip_max_norm: 0.1 # gradient clipping

  batch_size: 8 # physical, should fit on the device. In DDP used per GPU
  b_accum_steps: 1 # grad accumulation (n * bs)
  epochs: 100
  early_stopping: 30 # 0 - no early stopping
  ignore_background_epochs: 0 # background images are not used for N epochs in train set
  num_workers: 8 # In DDP used per GPU
  mask_batch_size: 1000 # number of images to process at once when computing mask metrics

  ### Validation ###
  conf_thresh: 0.5
  iou_thresh: 0.5

  ### EMA ###
  use_ema: True # use exponential moving average model
  ema_momentum: 0.9998

  ### Optimizer and Scheduler ###
  base_lr: ${train.lrs.${model_name}.base_lr}
  backbone_lr: ${train.lrs.${model_name}.backbone_lr}
  cycler_pct_start: 0.1
  weight_decay: 0.000125
  betas: [0.9, 0.999]
  label_smoothing: 0.0

  ### Augs ###
  mosaic_augs: # not recommended for segmentation
    mosaic_prob: 0 # set to 0 to disable
    no_mosaic_epochs: 5
    mosaic_scale: [0.5, 1.5]
    degrees: 0.0 # not recommended if bbox precision is critical
    translate: 0.2
    shear: 2.0

  augs:
    rotation_degree: 10 # maximum +- rotation
    rotation_p: 0.0 # probability of the rotation (with above degree)
    multiscale_prob: 0.0
    rotate_90: 0.05
    left_right_flip: 0.3
    up_down_flip: 0.0
    to_gray: 0.01
    blur: 0.01
    gamma: 0.02
    brightness: 0.02
    noise: 0.01
    coarse_dropout: 0.0

  ### Reproducibility ###
  seed: 42
  cudnn_fixed: False

  ### Recommended learning rates ###
  lrs:
    n:
      backbone_lr: 0.0004
      base_lr: 0.0008
    s:
      backbone_lr: 0.00006  # can setup up to 0.0002
      base_lr: 0.00025  # 0.0004
    m:
      backbone_lr: 0.00002  # 0.000025
      base_lr: 0.00015  # 0.00025
    l:
      backbone_lr: 0.00000625 # 0.0000125
      base_lr: 0.000125 # 0.00025
    x:
      backbone_lr: 0.0000015  # 0.0000025
      base_lr: 0.0001  # 0.00025


split:
  ignore_negatives: False # only use images with labels
  shuffle: True
  train_split: 0.85
  val_split: 0.15 # test_split = 1 - train_split - val_split


export: # TensorRT must be done on the inference device
  half: True # tensorrt, openvino
  max_batch_size: 1 # torch, tensorrt, openvino
  dynamic_input: False
  ov_int8_max_drop: 0.01


infer:
  to_crop: True  # if True - saves crops of detected objects
  paddings: # if int - amount of pixes, if float - percentage of image size
    w: 0.05
    h: 0.05


### service ###
defaults:
  - _self_
  - override hydra/hydra_logging: disabled
  - override hydra/job_logging: disabled

hydra:
  output_subdir: null
  run:
    dir: .

now_dir: &nowdir ${now:%Y-%m-%d}
```

**Comment by ArgoHA** [comment]
Hey, I pushed another improvement and tested on your dataset, here is the config:

```
project_name: sam_44_mss # for wandb
exp_name: sam_44_mss # experiment name

exp: ${exp_name}_${now_dir}

model_name: s # model size (n, s, m, l, x)
task: segment

train:
  ### Paths ###
  root: /home/incognito/D-FINE-seg # project root with dataset and outputs
  pretrained_dataset: coco # coco, obj2coco
  pretrained_model_path: pretrained/dfine_${model_name}_${train.pretrained_dataset}.pth # dfine_m_obj2coco.pth

  data_path: ${train.root}/datasets/sam_44_mss # path to dataset
  path_to_test_data: ${train.root}/images  # path to test set, used in infer script
  path_to_save: ${train.root}/output/models/${exp} # where to save output

  debug_img_path: ${train.root}/output/debug_images
  eval_preds_path: ${train.root}/output/eval_preds
  bench_img_path: ${train.root}/output/bench_imgs
  infer_path: ${train.root}/output/infer

  ### Configs ###
  use_wandb: False
  device: cuda
  label_to_name: # dataset's classes
    0: textline

  use_one_class: False

  ### Distributed Training (DDP) ###
  ddp:
    enabled: False  # set to true to use multi-GPU
    n_gpus: 2  # how many GPUs / processes per node for torchrun

  decision_metrics:
  - f1
  # - mAP_50

  img_size: [640, 640] # (h, w)
  keep_ratio: False # image aspect ratio, if True - image will be padded
  to_visualize_eval: True # save images with gt and preds
  debug_img_processing: True # save images after preprocessing

  amp_enabled: True # use automatic mixed precision
  clip_max_norm: 0.1 # gradient clipping

  batch_size: 8 # physical, should fit on the device. In DDP used per GPU
  b_accum_steps: 1 # grad accumulation (n * bs)
  epochs: 100
  early_stopping: 30 # 0 - no early stopping
  ignore_background_epochs: 0 # background images are not used for N epochs in train set
  num_workers: 8 # In DDP used per GPU
  mask_batch_size: 1000 # number of images to process at once when computing mask metrics

  ### Validation ###
  conf_thresh: 0.5
  iou_thresh: 0.5

  ### EMA ###
  use_ema: True # use exponential moving average model
  ema_momentum: 0.9998

  ### Optimizer and Scheduler ###
  use_scheduler: True
  base_lr: ${train.lrs.${model_name}.base_lr}
  backbone_lr: ${train.lrs.${model_name}.backbone_lr}
  cycler_pct_start: 0.1
  weight_decay: 0.000125
  betas: [0.9, 0.999]
  label_smoothing: 0.0

  ### Augs ###
  mosaic_augs: # not recommended for segmentation
    mosaic_prob: 0.0 # set to 0 to disable
    no_mosaic_epochs: 5
    mosaic_scale: [0.5, 1.5]
    degrees: 0.0 # not recommended if bbox precision is critical
    translate: 0.2
    shear: 2.0

  augs:
    rotation_degree: 10 # maximum +- rotation
    rotation_p: 0.0 # probability of the rotation (with above degree)
    multiscale_prob: 0.0
    rotate_90: 0.05
    left_right_flip: 0.3
    up_down_flip: 0.0
    to_gray: 0.01
    blur: 0.01
    gamma: 0.02
    brightness: 0.02
    noise: 0.01
    coarse_dropout: 0.0

  ### Reproducibility ###
  seed: 42
  cudnn_fixed: False

  ### Recommended learning rates ###
  lrs:
    n:
      backbone_lr: 0.0004
      base_lr: 0.0008
    s:
      backbone_lr: 0.00006  # can setup up to 0.0002
      base_lr: 0.00025  # 0.0004
    m:
      backbone_lr: 0.00002  # 0.000025
      base_lr: 0.00015  # 0.00025
    l:
      backbone_lr: 0.00000625 # 0.0000125
      base_lr: 0.000125 # 0.00025
    x:
      backbone_lr: 0.0000015  # 0.0000025
      base_lr: 0.0001  # 0.00025


split:
  ignore_negatives: False # only use images with labels
  shuffle: True
  train_split: 0.85
  val_split: 0.15 # test_split = 1 - train_split - val_split


export: # TensorRT must be done on the inference device
  half: True # tensorrt, openvino
  max_batch_size: 1 # torch, tensorrt, openvino
  dynamic_input: False
  ov_int8_max_drop: 0.01


infer:
  to_crop: True  # if True - saves crops of detected objects
  paddings: # if int - amount of pixes, if float - percentage of image size
    w: 0.05
    h: 0.05


### service ###
defaults:
  - _self_
  - override hydra/hydra_logging: disabled
  - override hydra/job_logging: disabled

hydra:
  output_subdir: null
  run:
    dir: .

now_dir: &nowdir ${now:%Y-%m-%d}
```

Looks fine even after 13 epochs

<img width="956" height="133" alt="Image" src="https://github.com/user-attachments/assets/974c1805-a02d-4eca-978c-aa6a60caf7cb" />

**Comment by johnlockejrr** [comment]
> Looks good to me. Can you compare results to ultralytics and rf-detr? How is it overall for you?
> 
> Btw, nice mask visualization with that opacity and edges, I will probably do something similar when I have time for it

Sure, I'll do that. With YOLO I always had lost lines and edges not detected.

**Comment by johnlockejrr** [comment]
I will make a test between D-FINE-seg, YOLO, RF-DETR.
Anyway, here is an example I get from RF-DETR:

![Image](https://github.com/user-attachments/assets/6ca22f3d-40f1-4eba-9f06-93806886f623)

**Comment by johnlockejrr** [comment]
YOLO11 seg same dataset S model trained for 308 epochs.

![Image](https://github.com/user-attachments/assets/c71085be-86f8-4026-9faa-67941019c2f4)

**Comment by ArgoHA** [comment]
You can do that, but scheduler and EMA will be reset, so it's not exactly the same as let's say you just trained for longer 1 time. Currently resume is not fully supported. But starting from prev weights should be ok

**Comment by johnlockejrr** [comment]
Very good results training on square images of `640×640` but would be a bad idea to train on `640×896` or even `768×1024` sizes keeping in mind my task is for manuscript pages where normally they are not square? Or, after pretraining on synthetic data on `640×640` size, would be a bad idea to finetune on real ground-truth for another page size like `640×896` for example?

```
2025-12-31 12:06:44.410 | INFO     | __main__:save_model:483 - Saving new best model🔥
2025-12-31 12:06:44.630 | INFO     | src.dl.utils:log_metrics_locally:92 - Metrics on epoch 24:
+-----+--------+-------------+--------+-----------+--------+--------+---------+-------+------+
|     | mAP_50 | mAP_50_mask |   f1   | precision | recall |  iou   |   TPs   |  FPs  | FNs  |
+-----+--------+-------------+--------+-----------+--------+--------+---------+-------+------+
| val | 0.9891 |   0.9885    | 0.9921 |  0.9885   | 0.9956 | 0.7642 | 10359.0 | 120.0 | 46.0 |
+-----+--------+-------------+--------+-----------+--------+--------+---------+-------+------+
```

**Comment by ArgoHA** [comment]
It should be fine to train in the size that fits your task. I would pretrain and finetune in the same manner. And for sure you should run an experiment and compare results. In theory bigger input size + less warping should be good, but model was pretrained on 640x640 (coco) so positional encoder might be a little biased to squares.

So, it's bot a bad idea, but decide based on your metrics

**Comment by ArgoHA** [comment]
I think it will take ~80 hours to fully train even `s` model. But the problem is that with coco and training from scratch default parameters might not work (I had to lower the lr, turn off scheduler and still failed at epoch 77 because of blown gradients). 

Also because backbone, encoder and detection decoder are pretrained - pretraining from scratch with Mask Decoder might not even do much. And I don't think it will make any difference in your task.

You can try this [preptrained model on coco masks](https://drive.google.com/file/d/1dlOQw7h1ie_14QMwfnqlcnDjBy_Icn-v/view?usp=share_link). If it doesn't make a difference for you (to train on your dataset initing these weights instead of original coco), then it doesn't make sense for you to pretrain coco masks

**Comment by johnlockejrr** [comment]
In my use case, training from detection models with synthetic data then real data works very well. Right now I'm training on non square `896x640` and compare the results.
Even on 6th epoch looks great already:

<img width="1161" height="178" alt="Image" src="https://github.com/user-attachments/assets/be2aaf11-bae0-43a2-9046-fb6554d51dcc" />

**Comment by ArgoHA** [comment]
Perfect, that's great to hear. Thanks for testing things out so quickly

**Comment by ArgoHA** [comment]
Yeah, I am considering that, thank you

**Comment by johnlockejrr** [comment]
Is the DDP tested/ready?

<img width="1895" height="775" alt="Image" src="https://github.com/user-attachments/assets/0fcafc91-1288-4937-acfc-feb7eda0d4a8" />

**Comment by ArgoHA** [comment]
Yes, it was working last time I tried. You need to run `make train` with ddp enabled in configs

**Comment by johnlockejrr** [comment]
Right now, training on a very complicated dataset and the training handles it very well.

<img width="1897" height="230" alt="Image" src="https://github.com/user-attachments/assets/2d8b0a18-9ee9-42cd-b8f0-7228df6cb855" />

For data like this:

[Bergskollegium__Relationer_och_skrivelser__E3_7__1698__40004028_00015.tif](https://github.com/user-attachments/files/24420641/Bergskollegium__Relationer_och_skrivelser__E3_7__1698__40004028_00015.tif)

![Image](https://github.com/user-attachments/assets/b49aefb0-bf8b-4f6b-be45-d2f2452cb84e)

**Comment by johnlockejrr** [comment]
Any idea what could cause this catastrophic evaluation collapse?

<img width="1897" height="966" alt="Image" src="https://github.com/user-attachments/assets/abc8d59d-b5bc-4baf-9cb2-23329c228e36" />
